common/src/test/java/io/netty/util/ResourceLeakDetectorTest.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util;
+
+import org.junit.Test;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+import java.util.concurrent.BrokenBarrierException;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class ResourceLeakDetectorTest {
+
+    @Test(timeout = 30000)
+    public void testConcurentUsage() throws Throwable {
+        final AtomicBoolean finished = new AtomicBoolean();
+        final AtomicReference<Throwable> error = new AtomicReference<Throwable>();
+        // With 50 threads its reproducable on every run.
+        Thread[] threads = new Thread[50];
+        final CyclicBarrier barrier = new CyclicBarrier(threads.length);
+        for (int i = 0; i < threads.length; i++) {
+            Thread t = new Thread(new Runnable() {
+                Queue<LeakAwareResource> resources = new ArrayDeque<LeakAwareResource>(100);
+
+                @Override
+                public void run() {
+                    try {
+                        barrier.await();
+
+                        // Run 10000 times or until the test is marked as finished.
+                        for (int b = 0; b < 1000 && !finished.get(); b++) {
+
+                            // Allocate 100 LeakAwareResource per run and close them after it.
+                            for (int a = 0; a < 100; a++) {
+                                DefaultResource resource = new DefaultResource();
+                                ResourceLeak leak = DefaultResource.detector.open(resource);
+                                LeakAwareResource leakAwareResource = new LeakAwareResource(resource, leak);
+                                resources.add(leakAwareResource);
+                            }
+                            if (closeResources(true)) {
+                                finished.set(true);
+                            }
+                        }
+                    } catch (BrokenBarrierException e) {
```

Comment:
should we `catch Throwable` after the other exceptions (e.g. `InterruptedException`) to make sure we bubble up the exception and it is visible from the junit thread?